client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -203,8 +214,41 @@ private Checkpoint completeCheckpoint(String checkpointName) {
 
     @Override
     public void resetReaderGroup(ReaderGroupConfig config) {
-        Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, config);
-        synchronizer.updateStateUnconditionally(new ReaderGroupStateInit(config, segments, getEndSegmentsForStreams(config)));
+        while (true) {
+            synchronizer.fetchUpdates();
+            val currentConfig = synchronizer.getState().getConfig();
+            // We only move into the block if the state transition has happened successfully.
+            if (stateTransition(currentConfig, new UpdatingConfig(true))) {
+                // Use the latest generation and reader group Id.
+                ReaderGroupConfig newConfig = config.toBuilder()
+                        .readerGroupId(currentConfig.getReaderGroupId())
+                        .generation(currentConfig.getGeneration()).build();
+                boolean success = Futures.getThrowingException(controller.updateReaderGroup(scope, groupName, newConfig));
+                Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, newConfig);
```

Comment:
`getSegmentsForStreams`  should be inside the `if (success)` block since it is not needed if controller.updateReaderGroup() has failed.

Reply:
It is inside the if block